generator/cells_merger.cpp

```diff
@@ -0,0 +1,273 @@
+                        TryGet(startXy.x + 1, startXy.y + 1).GetSum();
+  directions[0] = {topRight, m2::PointI{startXy.x + 1, startXy.y + 1}};
+  auto const topLeft = TryGet(startXy.x - 1, startXy.y).GetSum() +
+                       TryGet(startXy.x, startXy.y + 1).GetSum() +
+                       TryGet(startXy.x - 1, startXy.y + 1).GetSum();
+  directions[1] = {topLeft, m2::PointI{startXy.x - 1, startXy.y + 1}};
+  auto const bottomRight = TryGet(startXy.x + 1, startXy.y).GetSum() +
+                           TryGet(startXy.x, startXy.y - 1).GetSum() +
+                           TryGet(startXy.x + 1, startXy.y - 1).GetSum();
+  directions[2] = {bottomRight, m2::PointI{startXy.x + 1, startXy.y - 1}};
+  auto const bottomLeft = TryGet(startXy.x - 1, startXy.y).GetSum() +
+                          TryGet(startXy.x, startXy.y - 1).GetSum() +
+                          TryGet(startXy.x - 1, startXy.y - 1).GetSum();
+  directions[3] = {bottomLeft, m2::PointI{startXy.x - 1, startXy.y - 1}};
+  auto const nextXy = std::max_element(std::cbegin(directions), std::cend(directions))->second;
+  auto const direction = nextXy - startXy;
+  return Has(startXy.x + direction.x, startXy.y + direction.y) ? direction : m2::PointI();
+}
+
+void CellsMerger::Remove(m2::PointI const & minXy, m2::PointI const & maxXy)
+{
+  auto const xMin = std::min(minXy.x, maxXy.x);
+  auto const xMax = std::max(minXy.x, maxXy.x);
+  auto const yMin = std::min(minXy.y, maxXy.y);
+  auto const yMax = std::max(minXy.y, maxXy.y);
+  for (int32_t x = xMin; x <= xMax; ++x)
+  {
+    for (int32_t y = yMin; y <= yMax; ++y)
+      m_matrix.erase({x, y});
+  }
+}
+
+bool CellsMerger::Has(int32_t x, int32_t y) const
+{
+  return m_matrix.count({x, y}) != 0;
+}
+
+bool CellsMerger::Has(const m2::PointI & xy) const
+{
+  return Has(xy.x, xy.y);
+}
+
+m2::PointI CellsMerger::FindMax() const
+{
+  m2::PointI max(-1, -1);
+  size_t sum = 0;
+  for (auto const & pair : m_matrix)
+  {
+    auto const cellSum = pair.second.GetSum();
+    if (cellSum > sum)
+    {
+      sum = cellSum;
+      max = pair.first;
+    }
+  }
+  return max;
+}
+
+m2::RectD CellsMerger::Union(m2::PointI const & startXy)
+{
+  m2::PointI nextXy = startXy;
+  auto const direction = FindDirection(startXy);
+  m2::RectD united;
+  if (direction.x == 0)
+  {
+    united = Get(startXy).GetCell();
+  }
+  else
+  {
+    while (true)
+    {
+      m2::PointI temp(nextXy.x + direction.x, nextXy.y + direction.y);
+      if (!MayTransform(startXy, temp))
+        break;
+      nextXy = temp;
+    }
+
+    auto cell1 = Get(startXy).GetCell();
+    auto const & cell2 = Get(nextXy).GetCell();
+    cell1.Add(cell2);
+    united = cell1;
+  }
+  Remove(startXy, nextXy);
+  return united;
+}
+
+std::vector<m2::RectD> MakeNet(double step, double minX, double minY, double maxX, double maxY)
+{
+  std::vector<m2::RectD> net;
+  auto xmin = minX;
+  auto ymin = minY;
+  auto x = xmin;
+  auto y = ymin;
+  while (xmin < maxX)
```

Comment:
CHECK that min <= max before starting, please

Reply:
++